### PR TITLE
Removing promise wrapper and just using native promise, issue #21

### PR DESCRIPTION
**Pull request by rorticus**
**Type:** feature

**Description:** 

Removing the promise wrapper and just using the native Promise object, if available.  Note that by doing this we had to remove `finally` because it is not a part of the native promises.

Also, it appears that the old promise wrapper was suppressing all uncaught promise errors. I don't know if this was intentional or not, but the `copy` method had a `this.then(fn, fn)` in it, which would swallow everything

**Related Issue:** #21 

Please review this checklist before submitting your PR:
- [x] There is a related issue
- [x] All contributors have signed a CLA
- [x] All code matches the [style guide](https://github.com/dojo/meta/blob/master/STYLE.md)
- [x] The code passes the CI tests
- [x] Unit or Functional tests are included in the PR
- [x] The PR increases or maintains the overall unit test coverage percentage
- [x] The code is ready to be merged

### CODE REVIEW

File: src/Promise.ts [hunk 1]
```diff
@@ -48,437 +27,354 @@ export function isThenable<T>(value: any): value is Thenable<T> {
...
 	 * }).then((value) => {
 	 *     value.foo === 'foo'; // true
 	 *     value.bar === 'bar'; // true
 	 * });
 	 */
-	static all<T>(iterable: Iterable<(T | Thenable<T>)> | (T | Thenable<T>)[]): Promise<T[]> {
-		return this.copy(Promise.PromiseConstructor.all(unwrapPromises(iterable)));
-	}
+	public static all<T>(iterable: Iterable<(T | Thenable<T>)> | (T | Thenable<T>)[]): Promise<T[]> {
+		throw new Error();
```
Comment (kitsonk) [inline src/Promise.ts:320]: We should add instanbul ignore here.

File: src/Promise.ts [hunk 2]
```diff
@@ -48,437 +27,354 @@ export function isThenable<T>(value: any): value is Thenable<T> {
...
+	 *     foo: Promise.resolve('foo'),
+	 *     bar: Promise.resolve('bar')
 	 * }).then((value) => {
 	 *     value === 'foo'; // true
 	 * });
 	 */
-	static race<T>(iterable: Iterable<(T | Thenable<T>)> |  (T | Thenable<T>)[]): Promise<T> {
-		return this.copy(Promise.PromiseConstructor.race(unwrapPromises(iterable)));
+	public static race<T>(iterable: Iterable<(T | Thenable<T>)> |  (T | Thenable<T>)[]): Promise<T> {
+		throw new Error();
```
Comment (kitsonk) [inline src/Promise.ts:343]: Same here....

File: src/Promise.ts [hunk 3]
```diff
@@ -48,437 +27,354 @@ export function isThenable<T>(value: any): value is Thenable<T> {
...
-	 */
-	static resolve(): Promise<void>;
-	static resolve<T>(value: (T | Thenable<T>)): Promise<T>;
-	static resolve<T>(value?: any): Promise<T> {
-		if (value instanceof Promise) {
-			return value;
-		}
-		return this.copy(Promise.PromiseConstructor.resolve(value));
+	public static reject<T>(reason?: Error): Promise<any> {
+		throw new Error();
```
Comment (kitsonk) [inline src/Promise.ts:351]: and again

File: src/Promise.ts [hunk 4]
```diff
@@ -48,437 +27,354 @@ export function isThenable<T>(value: any): value is Thenable<T> {
...
-
-		this._state = State.Pending;
-		this.promise.then(
-			() => { this._state = State.Fulfilled; },
-			() => { this._state = State.Rejected; }
-		);
+	public static resolve(): Promise<void>;
+	public static resolve<T>(value: (T | Thenable<T>)): Promise<T>;
+	public static resolve<T>(value?: any): Promise<T> {
+		throw new Error();
```
Comment (kitsonk) [inline src/Promise.ts:361]: and again

File: src/Promise.ts [hunk 5]
```diff
@@ -48,437 +27,354 @@ export function isThenable<T>(value: any): value is Thenable<T> {
...
-	}
-
-	/**
-	 * The current Promise state.
-	 */
-	get state(): State {
-		return this._state;
+	public catch<U>(onRejected: (reason: Error) => (U | Thenable<U>)): Promise<U>;
+	public catch<U>(onRejected: (reason: Error) => void): Promise<U> {
+		throw new Error();
```
Comment (kitsonk) [inline src/Promise.ts:370]: again...

File: src/Promise.ts [hunk 6]
```diff
@@ -48,437 +27,354 @@ export function isThenable<T>(value: any): value is Thenable<T> {
...
 
-	/**
-	 * Adds a callback to the promise to be invoked when the asynchronous operation completes successfully.
-	 */
-	then<U>(onFulfilled?: ((value: T) => (U | Thenable<U> | null | undefined)) | null | undefined, onRejected?: (reason: Error) => void): Promise<U>;
-	then<U>(onFulfilled?: ((value: T) => (U | Thenable<U> | null | undefined)) | null | undefined, onRejected?: (reason: Error) => (U | Thenable<U>)): Promise<U> {
-		return (<typeof Promise> this.constructor).copy(this.promise.then(onFulfilled, onRejected));
+	public then<U>(onFulfilled?: ((value: T) => (U | Thenable<U> | null | undefined)) | null | undefined, onRejected?: (reason: Error) => void): Promise<U>;
+	public then<U>(onFulfilled?: ((value: T) => (U | Thenable<U> | null | undefined)) | null | undefined, onRejected?: (reason: Error) => (U | Thenable<U>)): Promise<U> {
+		throw new Error();
```
Comment (kitsonk) [inline src/Promise.ts:376]: ok, one more time

**Review by kitsonk — COMMENTED** [review]
Initial review, will be doing more later.

**Comment by rorticus** [comment]
Fixed all that istanbul stuff :)

**Comment by rorticus** [comment]
Updated to match style guide and increased test coverage.

**Comment by kitsonk** [comment]
Ok, I have linked this into `dojo-core` and then subsequently linked it into `dojo-compose` and `dojo-widgets` and not encountered any issues.  There is still a risk that with something as core as `Promise` we might have challenges, but I think this is good to go now.

**Review by kitsonk — APPROVED** [review]
👍 🎃

**Pull request merged**